smart-engine/src/main/java/org/smartdata/server/engine/rule/SmallFilePlugin.java

```diff
@@ -0,0 +1,188 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.smartdata.server.engine.rule;
+
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.smartdata.hdfs.action.HdfsAction;
+import org.smartdata.hdfs.action.SmallFileCompactAction;
+import org.smartdata.metastore.MetaStore;
+import org.smartdata.metastore.MetaStoreException;
+import org.smartdata.model.CmdletDescriptor;
+import org.smartdata.model.FileInfo;
+import org.smartdata.model.FileState;
+import org.smartdata.model.RuleInfo;
+import org.smartdata.model.rule.RuleExecutorPlugin;
+import org.smartdata.model.rule.TranslateResult;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+public class SmallFilePlugin implements RuleExecutorPlugin {
+  private int batchSize;
+  private MetaStore metaStore;
+  private String containerFileDir;
+  private static final String CONTAINER_FILE_PREFIX = "container_file_";
+  private static final Logger LOG = LoggerFactory.getLogger(SmallFilePlugin.class);
+
+  public SmallFilePlugin(MetaStore metaStore, String containerFileDir, int batchSize) {
+    this.metaStore = metaStore;
+    this.batchSize = batchSize;
+    if (containerFileDir.charAt(containerFileDir.length() - 1) != '/') {
+      this.containerFileDir = containerFileDir + "/";
+    } else {
+      this.containerFileDir = containerFileDir;
+    }
+  }
+
+  @Override
+  public void onNewRuleExecutor(final RuleInfo ruleInfo, TranslateResult tResult) {}
+
+  @Override
+  public boolean preExecution(final RuleInfo ruleInfo, TranslateResult tResult) {
+    return true;
+  }
+
+  @Override
+  public List<String> preSubmitCmdlet(final RuleInfo ruleInfo, List<String> objects) {
+    if (ruleInfo.getRuleText().contains("compact")) {
+      // Split valid small files according to the file permission
+      Map<FilePermission, List<String>> filePermissionMap = new HashMap<>(32);
+      try {
+        List<String> validFiles = getValidFileList(objects);
+        for (String element : validFiles) {
+          FilePermission filePermission = new FilePermission(metaStore.getFile(element));
```

Comment:
Merge with `getValidFileList` to reduce the number of calls to `metaStore.getFile`